src/ray/raylet/node_manager.cc

```diff
@@ -1633,11 +1673,42 @@ void NodeManager::HandleReturnWorker(const rpc::ReturnWorkerRequest &request,
   std::shared_ptr<Worker> worker = std::move(leased_workers_[worker_id]);
 
   if (new_scheduler_enabled_) {
+    if (worker->IsBlocked()) {
+      // If worker blocked, unblock it to return the cpu resources back to the worker.
+      HandleDirectCallTaskUnblocked(worker);
+    }
     auto it = leased_worker_resources_.find(worker_id);
     RAY_CHECK(it != leased_worker_resources_.end());
-    new_resource_scheduler_->AddNodeAvailableResources(client_id_.Binary(), it->second);
+
+    new_resource_scheduler_->AddNodeAvailableResources(client_id_.Binary(),
+      it->second.GetResourceMap());
+
+    if (worker->borrowed_cpu_resources_.GetResourceMap().size()) {
+      // This machine is oversubscribed, so the worker didn't get back cpus when unblocked.
+      // Thus we need to substract these cpus, as the previous "AddNodeAvailableResources" call
+      // assumed they were allocated to this worker.
+      new_resource_scheduler_->SubtractNodeAvailableResources(client_id_.Binary(),
+          worker->borrowed_cpu_resources_.GetResourceMap());
+      worker->borrowed_cpu_resources_ = ResourceSet();
+    }
     leased_worker_resources_.erase(it);
+
+    // Update resource ids
+    auto const &task_resources = worker->GetTaskResourceIds();
+      const ClientID &client_id = gcs_client_->client_table().GetLocalClientId();
+    local_available_resources_.ReleaseConstrained(
+        task_resources, cluster_resource_map_[client_id].GetTotalResources());
+    cluster_resource_map_[gcs_client_->client_table().GetLocalClientId()].Release(
+        task_resources.ToResourceSet());
+    worker->ResetTaskResourceIds();
+
     NewSchedulerSchedulePendingTasks();
+
+    // TODO (ion): Handle ProcessDisconnectClientMessage()
```

Comment:
This seems important. What happens if we don't handle it? Will it be done in this PR?

Reply:
This will be done in the next PR. This PR was already more complex than anticipated, because handling resource ids. Again, the new scheduler won't be enabled in this PR.